### PR TITLE
Add support for wildcard suffix in WildcardIndex #30413

### LINKED ISSUE #30413: large-message-destinations not supported for response to ask
[Large message destinations](https://doc.akka.io/docs/akka/current/remoting-artery.html#dedicated-subchannel-for-large-messages) can be configured to send larger messages (by default >256KB and <2MB) to remote actors. This does not work in combination with ClusterSharding.

I've created a [minimal test case](https://github.com/an-tex/akka-samples/blob/2.6/akka-sample-cluster-scala/src/multi-jvm/scala/sample/cluster/stats/LargeMessageSpec.scala) reproducing the issue using a multi jvm test with a three node cluster. It's based on this [existing test case](https://github.com/akka/akka-samples/blob/2.6/akka-sample-cluster-scala/src/multi-jvm/scala/sample/cluster/stats/StatsSampleSpec.scala).

Inside this repo run `sbt multi-jvm:testOnly *.LargeMessageSpec` and you'll see the error

`[JVM-3] akka.remote.OversizedPayloadException: Discarding oversized payload sent to Some(Actor[akka://LargeMessageSpec@127.0.0.1:61038/system/sharding/largeMessageTypeKey#-806848135]): max allowed size 262144 bytes. Message type [akka.cluster.sharding.typed.ShardingEnvelope].`

even though there's a [large destination configured](https://github.com/an-tex/akka-samples/blob/2.6/akka-sample-cluster-scala/src/main/resources/application.conf#L23) like this:

```
akka.remote.artery.large-message-destinations = [
  "/system/singletonManagerlargeMessageSingleton/*",
  "/system/sharding/largeMessageTypeKey*"
]
```

Note: This error occurs only on two of the nodes as serialization occurs only between JVM boundaries 

To make sure the `large-message-destinations` are actually working at all there's a separate test case using a ClusterSingleton where no issues occur.

Versions:
```scala
val akkaVersion = "2.6.14"
scalaVersion := "2.13.5"
```

### PR DESCRIPTION
**Pull request by an-tex**
Resolves #30413

Please consider this a draft. I'd like to get feedback on the implementation especially around the points @johanandren mentioned in https://github.com/akka/akka/issues/30413#issuecomment-934369333 . Once that's all figured out I'd also update the documentation.

A few things to note:

As far as I can see the WildCardIndex is used in three places:
- [`largeMessageDestinations: WildcardIndex[NotUsed]`](https://github.com/akka/akka/blob/main/akka-remote/src/main/scala/akka/remote/artery/Association.scala#L140)
- [`priorityMessageDestinations: WildcardIndex[NotUsed]`](https://github.com/akka/akka/blob/main/akka-remote/src/main/scala/akka/remote/artery/Association.scala#L141)
- [`private val deployments = new AtomicReference(WildcardIndex[Deploy]())`](https://github.com/akka/akka/blob/main/akka-actor/src/main/scala/akka/actor/Deployer.scala#L220)

There shouldn't be any performance implications when there are no wildcard suffixes (only an additional find operation on an empty Map)

I've added a test case to the `LargeMessagesSpec`. It feels like it doesn't quite belong there but as it verifies the original issue it might be still worth having it. Additionally I've used the [reproducing test case from the original issue](https://github.com/an-tex/akka-samples/blob/2.6/akka-sample-cluster-scala/src/multi-jvm/scala/sample/cluster/stats/LargeMessageSpec.scala) to manually verify this fix in a more end-2-end scenario.

The bincomp issues around the [WildcardTree](https://github.com/akka/akka/blob/main/akka-actor/src/main/scala/akka/util/WildcardIndex.scala#L45) shouldn't be an issue as the `WildcardTree` is only used within [WildcardIndex](https://github.com/akka/akka/blob/main/akka-actor/src/main/scala/akka/util/WildcardIndex.scala#L10) which signatures haven't changed

### CODE REVIEW

File: akka-actor/src/main/scala/akka/util/WildcardIndex.scala [hunk 1]
```diff
@@ -53,18 +54,31 @@ private[akka] final case class WildcardTree[T](
       copy(data = Some(d))
     } else {
       val e = elems.next()
-      copy(children = children.updated(e, children.getOrElse(e, WildcardTree[T]()).insert(elems, d)))
+      if (e != "*" && e != "**" && e.endsWith("*"))
+        copy(
+          wildcardSuffixChildren = wildcardSuffixChildren
+            .updated(e.stripSuffix("*"), wildcardSuffixChildren.getOrElse(e, WildcardTree[T]()).insert(elems, d)))
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/util/WildcardIndex.scala:60]: Should it perhaps fail for `e != "**" && e.endsWith("**")` for extra clarity that doesn't work?

**Review by johanandren — COMMENTED** [review]
Ah, it turns out I was wrong about actor selection, that has its own tree pattern match impl. good, then I think this is a good way forward.

I think it's good with the additional test coverage in the LargeMessageStreamSpec

**Comment by johanandren** [comment]
Current CI failure is formatting and binary compatibility check failure.

For the formatting please re-compile the tests (or run scalafmtAll manually) and check in the formatted sources.
For the binary compatibility it is safe to filter those warnings since this is an internal class, no public API, see https://github.com/akka/akka/blob/main/CONTRIBUTING.md#binary-compatibility for details about what file and where etc.

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

**Comment by an-tex** [comment]
hey @johanandren, would you mind pinging some other reviewer? would be great to have it in for the next release :)

**Review by patriknw — APPROVED** [review]
LGTM, thanks for contributing

**Pull request merged**